msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/AcquireTokenSilentSupplier.java

```diff
@@ -120,4 +82,74 @@ AuthenticationResult execute() throws Exception {
 
         return res;
     }
+
+    private AuthenticationResult makeRefreshRequest(AuthenticationResult cachedResult,  Authority requestAuthority) throws Exception {
+        RefreshTokenRequest refreshTokenRequest = new RefreshTokenRequest(
+                RefreshTokenParameters.builder(silentRequest.parameters().scopes(), cachedResult.refreshToken()).build(),
+                silentRequest.application(),
+                silentRequest.requestContext(),
+                silentRequest);
+
+        AcquireTokenByAuthorizationGrantSupplier acquireTokenByAuthorisationGrantSupplier =
+                new AcquireTokenByAuthorizationGrantSupplier(clientApplication, refreshTokenRequest, requestAuthority);
+
+        try {
+            AuthenticationResult refreshedResult = acquireTokenByAuthorisationGrantSupplier.execute();
+
+            refreshedResult.metadata().tokenSource(TokenSource.IDENTITY_PROVIDER);
+
+            log.info("Access token refreshed successfully.");
+            return refreshedResult;
+        } catch (MsalServiceException ex) {
+            //If the token refresh attempt threw a MsalServiceException but the refresh attempt was done
+            // only because of refreshOn, then simply return the existing cached token
+            if (!afterRefreshOn && (silentRequest.parameters().forceRefresh() || StringHelper.isBlank(cachedResult.accessToken()))) {
+                throw ex;
+            }
+            return cachedResult;
+        }
+    }
+
+    private void setRefreshTelemetry(AuthenticationResult cachedResult) {
```

Comment:
This duplicates some of logic in `shouldRefresh`. Can you make it more [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself), for example by having `shouldRefresh` return the enum value instead of boolean?

Also, it doesn't take into account claims.

Reply:
In the latest commit I moved this code into the shouldRefresh() method so that it also sets the telemetry when determining whether or not to refresh (and claims are treated just like a force refresh)